src/main/java/com/salesforce/cdp/queryservice/auth/TokenManager.java

```diff
@@ -0,0 +1,42 @@
+package com.salesforce.cdp.queryservice.auth;
+
+import com.salesforce.cdp.queryservice.util.TokenException;
+import okhttp3.OkHttpClient;
+
+import java.util.Optional;
+import java.util.Properties;
+
+public class TokenManager {
+
+    private final CoreTokenProvider tokenProvider;
+    private final TokenExchangeHelper tokenExchangeHelper;
+
+    private OffcoreToken offcoreToken = null;
+
+    public TokenManager(Properties properties, OkHttpClient client) throws TokenException {
+        tokenProvider = CoreTokenProviderFactory.getCoreTokenProvider(properties, client);
+        tokenExchangeHelper = new TokenExchangeHelper(properties, client);
+    }
+
+    public TokenManager(CoreTokenProvider tokenProvider, TokenExchangeHelper tokenExchangeHelper) {
+        this.tokenProvider = tokenProvider;
+        this.tokenExchangeHelper = tokenExchangeHelper;
+    }
+
+    public CoreToken getCoreToken() throws TokenException {
+        return tokenProvider.getCoreToken();
+    }
+
+    public OffcoreToken getOffcoreToken() throws TokenException {
+        if (tokenProvider instanceof OffcoreTokenProvider) {
+            return ((OffcoreTokenProvider) tokenProvider).getOffcoreToken();
```

Comment:
I am confused, tokenProvider is of type CoreTokenProvider . When would it match to OffcoreTokenProvider ? Can we keep `CoreTokenProvider` and `OffcoreTokenProvider` as child of `TokenProvider` and TokenProvider to have `getToken()` method only.

Reply:
Let me see what are the alternatives here.

Token Exchange is not really required in the providers and is following the same process everywhere.

Currently the OffcoreTokenProvider is implemented only for the Refresh Token based flow. This also follows the same method, but the only difference is the token is cached.

Reply:
Exposing a single getToken wouldn't help the usecase which I am trying to achieve, where the coretoken and offcore tokens need to be accessed

Reply:
Refactored, can you pls check again